src/vm/comutilnative.cpp

```diff
@@ -714,13 +714,75 @@ void QCALLTYPE Buffer::Clear(void *dst, size_t length)
     memset(dst, 0, length);
 }
 
+#define BULK_MOVE_WITH_WRITE_BARRIER_BLOCK_SIZE 0x8000
+
```

Comment:
I think it could be simpler if “chunking” happens on the managed side. Then it would be regular fcalls without polling and frames. (move itself does not trigger or throw)

Bulk w barrier move would just need to assert it is not asked to move too much at once.

Reply:
It may be simpler, but less efficient. It can either add a new frame that does the chunking (throughput overhead - for small block sizes in particular); or aggressive inline the chunking into every callsite to avoid the frame (code side overhead). Which tradeoff would you pick if the chunking happens on the managed side?

Reply:
I was thinking of something like:

```cs
ManagedHelperProbablyInlineable()
{
     // get refs, byte count and whether need to copy backwards
     nint count = . . .;
     if (count < CHUNK_SIZE)
     {
           NativeHelper(ref src, ref dst, count, backwards);
     }
     else
     {
           ManagedChunkingHelperDoNotInline(mt, ref src, ref dst, count, backwards);
     }
}

ManagedChunkingHelperDoNotInline(MethodTable *mt, ref byte src, ref byte dst, nint count, bool backwards)
{
       int elementSize = . . . ;
       while(count)
       {
              int chunk = Align(min(count, chunkSize), elementSize);
              NativeHelper(src, dst, chunk , backwards);
              count -= chunk;
              src = . . .;
              dst = . . .;
       }
}
```